File: apps/desktop/src/api/suppliers.ts
import { invoke } from "@tauri-apps/api/core";
import type { Supplier, SupplierInput } from "./types";

/**
 * Typed wrappers around the `commands::suppliers` Tauri commands. This is the only module that
 * should call `invoke()` for suppliers — components call these functions instead.
 *
 * Note: Tauri camelCases top-level command argument names for the JS side (e.g. Rust's
 * `include_inactive` becomes `includeInactive` here), but does NOT rename fields inside a struct
 * argument like `SupplierInput` — those keep their exact snake_case Rust field names since that
 * struct has no `#[serde(rename_all = ...)]`.
 */

export function listSuppliers(includeInactive: boolean): Promise<Supplier[]> {
  return invoke("list_suppliers", { includeInactive });
}

export function getSupplier(id: number): Promise<Supplier> {
  return invoke("get_supplier", { id });
}

export function createSupplier(input: SupplierInput): Promise<Supplier> {
  return invoke("create_supplier", { input });
}

export function updateSupplier(id: number, input: SupplierInput): Promise<Supplier> {
  return invoke("update_supplier", { id, input });
}

export function archiveSupplier(id: number): Promise<void> {
  return invoke("archive_supplier", { id });
}

export function reactivateSupplier(id: number): Promise<void> {
  return invoke("reactivate_supplier", { id });
}

export function deleteSupplier(id: number): Promise<void> {
  return invoke("delete_supplier", { id });
}

File: apps/desktop/src-tauri/src/commands/suppliers.rs
use tauri::State;

use crate::db::repositories::suppliers::{self, Supplier, SupplierInput};
use crate::error::{AppError, AppResult};
use crate::DbState;

#[tauri::command]
pub fn list_suppliers(db: State<DbState>, include_inactive: bool) -> AppResult<Vec<Supplier>> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    suppliers::list(&conn, include_inactive)
}

#[tauri::command]
pub fn get_supplier(db: State<DbState>, id: i64) -> AppResult<Supplier> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    suppliers::get(&conn, id)
}

#[tauri::command]
pub fn create_supplier(db: State<DbState>, input: SupplierInput) -> AppResult<Supplier> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    suppliers::create(&conn, input)
}

#[tauri::command]
pub fn update_supplier(db: State<DbState>, id: i64, input: SupplierInput) -> AppResult<Supplier> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    suppliers::update(&conn, id, input)
}

#[tauri::command]
pub fn archive_supplier(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    suppliers::set_active(&conn, id, false)
}

#[tauri::command]
pub fn reactivate_supplier(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    suppliers::set_active(&conn, id, true)
}

#[tauri::command]
pub fn delete_supplier(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    suppliers::delete(&conn, id)
}
